typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,916 @@
+
+```typescript
+type Email = string;
+```
+
+And also you are able to create types with properties.
+Symbol `?` here is used as a syntactic sugar for `string | undefined` type.
+```typescript
+type Document = {
+  name: string;
+  athor?: string;
+}
+```
+
+#### Newtype
+
+Creating a `newtype` is not so trivial.
+There is no such analog in TypeScript, but you can do it using tags.
+
+And here is how all this magic works.
+With intersection types (& syntax) you can add `tag` field for your type.
+Using `symbol` you are able to declare const-named properties on types.
+Word `unique` means that they are unique.
+Your type is tagged with unique symbol.
+
+```typescript
+type Email = string & { readonly __tag: unique symbol };
+type City = string & { readonly __tag: unique symbol };
+```
+
+Using `as` which means type assertion you are able to cast your base type to tagged type.
+And when you will try to pass value of another tagged type, compiler will say that two unique symbols are different.
+
+```typescript
+function sendMessage(email: Email);
+
+sendMessage("message"); // Argument of type 'string' is not assignable to parameter of type 'Email'.
+sendMessage("St. Petersburg" as City); // Argument of type 'City' is not assignable to parameter of type 'Email'.
+
+sendMessage("email@gmail.com" as Email); // Ok
+```
+
+Anyway, this is not an ideal solution for newtypes.
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
+And there is also a [library](https://github.com/sveseme/taghiro) which provide more pretty implementation for them.
+
+### Unit types, Unions, Data types, and Pattern Matching
+
+Haskell ADT is a constantly used functionality of language.
+It allows you to build your own big types from small blocks.
+And with pattern matching it is easy to get access to this data.
+
+Unfortunately there is no same way to build ADT, but we will start from the easiest enums and move to the more complex things showing what we can do using existing type system in TypeScript.
+
+#### Unit types
+
+Unit types are subtypes of primitive types that contain exactly one primitive value.
+You can use it like enums setting unique string values.
+Or even use different values types.
+
+```typescript
+type Result = true | "error" | 5;
+```
+
+You can pass it to the function and match by a concret type.
+
+```typescript
+const resultInterpreter = (result: Result): string => {
+  if (result === true) {
+    return "true";
+  }
+  if (result === "error") {
+    return result;
+  }
+  return "five";
+};
+```
+
+#### Unions
+
+Moving forward the next feature is Unions.
+Unions are not tagged, so it is just a condition of possible types.
+
+```typescript
+type Result = string | number | (() => string);
+```
+
+Here and in the example with Unit types if you already matched by some type compiler will know that it will not be present in the future.
```

Comment:
word "matched" implies that it was matched successfully, I would use "failed to match" in this paragraph.